crates/ott-balancer/src/discovery/dns.rs

```diff
@@ -0,0 +1,51 @@
+use async_trait::async_trait;
+use tracing::info;
+use trust_dns_resolver::TokioAsyncResolver;
+
+use super::*;
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct DnsDiscoveryConfig {
+    /// The port that monoliths should be listening on for load balancer connections.
+    pub monolith_port: u16,
+    pub dns_server: Option<String>,
+    pub query: String
+}
+
+pub struct DnsMonolithDiscoverer {
+    config: DnsDiscoveryConfig,
+}
+
+impl DnsMonolithDiscoverer {
+    pub fn new(config: DnsDiscoveryConfig) -> Self {
+        info!(
+            "Creating DockerMonolithDiscoverer, Docker DNS server: {:?}",
+            config.dns_server
+        );
+        let query = format!("{}", config.dns_server.clone().unwrap_or_default());
+        Self { config }
+    }
+}
+
+#[async_trait]
+impl MonolithDiscoverer for DnsMonolithDiscoverer {
+    async fn discover(&mut self) -> anyhow::Result<Vec<MonolithConnectionConfig>> {
+        let resolver =
+            TokioAsyncResolver::tokio_from_system_conf().expect("failed to create resolver");
```

Comment:
The case where `self.config.dns_server` is `Some` is not taken into account here.

Reply:
this comment will be resolved in Issue #1337